api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -130,15 +131,45 @@ private static Set<Integer> getIdsInternal(Type type, boolean includeStructIds)
     return visit(type, new GetProjectedIds(includeStructIds));
   }
 
+  private static Set<Integer> innerFilter(Set<Integer> filteredIds, Types.NestedField innerField) {
+    // If it's a struct and if all ids need to be filtered we remove all
+    if (innerField.type().isStructType()) {
+      if (innerField.type().asStructType().fields()
+              .stream().allMatch(f -> filteredIds.contains(f.fieldId()))) {
+        filteredIds.removeAll(innerField.type().asStructType().fields().stream()
+                .map(Types.NestedField::fieldId).collect(Collectors.toSet()));
+      }
+      innerField.type().asStructType().fields().forEach(f -> innerFilter(filteredIds, f));
+    }
+    return filteredIds;
+  }
+
+  private static void filter(Set<Integer> projectedIds, Types.StructType struct, Set<Integer> fieldIds,
+                                boolean filteredStruct) {
+    Set<Integer> filteredIds = Sets.newHashSet(fieldIds);
+    if (filteredStruct) {
+      struct.fields().forEach(f -> innerFilter(filteredIds, f));
+    }
+    projectedIds.removeAll(filteredIds);
+  }
+
   public static Types.StructType selectNot(Types.StructType struct, Set<Integer> fieldIds) {
+    return selectNot(struct, fieldIds, false);
+  }
+
+  public static Types.StructType selectNot(Types.StructType struct, Set<Integer> fieldIds, boolean filteredStruct) {
```

Comment:
Nit: This boolean name is a bit strange to me. Is there perhaps a more descriptive name for what this corresponds to? Maybe `doesHaveInnerStruct` or `doesHaveInnerProjectedStruct`?

I was going to ask you to inline the field name with the raw boolean values in the tests, like `true /* filteredStruct /*`, but that still left me with some confusion.

Reply:
Hi @kbendick,
Thanks for the review.
I had trouble to name this variable. It's may more doesNeedToFilteredInnerStruct. 
I add this variable because on some cases we need to filtered inner struct (ORC) in some cases we do not need.

Is doesNeedToFilteredInnerStruct is better ?

Reply:
i just renamed it to doesNeedToKeepInnerStruct
Because it's more if it's a struct and if all ids need to be filtered we need keep all. So we remove all ids from filtered ids set